sdk/core/core-client/test/storageChallengeAuthentication.spec.ts

```diff
@@ -0,0 +1,262 @@
+    await sendTestRequest();
+    await sendTestRequest();
+    await sendTestRequest();
+
+    assert.equal(getTokenStub.callCount, 1);
+  });
+
+  it("should refresh token if it has expired", async () => {
+    getTokenStub = sinon
+      .stub<[string | string[], GetTokenOptions | undefined], Promise<AccessToken | null>>()
+      .resolves({ expiresOnTimestamp: Date.now() - 1000000, token: "originalToken" });
+
+    const policy = bearerTokenAuthenticationPolicy({
+      credential: { getToken: getTokenStub },
+      scopes: ["https://example.org"],
+      challengeCallbacks: {
+        authorizeRequestOnChallenge: storageAuthorizeRequestOnChallenge,
+      },
+    });
+
+    const sendTestRequest = () =>
+      policy.sendRequest(
+        {
+          headers: createHttpHeaders(),
+          method: "GET",
+          requestId: "",
+          timeout: 1000,
+          url: "https://example.org",
+          withCredentials: true,
+        },
+        async (req) => {
+          assert.equal(req.headers.get("authorization"), "Bearer originalToken");
+          return {
+            headers: createHttpHeaders(),
+            request: req,
+            status: 200,
+          };
+        }
+      );
+
+    // Send the request 3 times
+    await sendTestRequest();
+    await sendTestRequest();
+    await sendTestRequest();
+
+    // We are setting expiration time in the past everytime we refresh the token
+    // so we expect 3 calls to refresh the token
+    assert.equal(getTokenStub.callCount, 3);
+  });
+
+  it("should refresh token if valid but within window it has expired", async () => {
+    getTokenStub = sinon
+      .stub<[string | string[], GetTokenOptions | undefined], Promise<AccessToken | null>>()
+      // Refresh window is 1000ms setting 999 to make sure we are in the refresh window
+      .resolves({ expiresOnTimestamp: Date.now() + 999, token: "originalToken" });
+
+    const policy = bearerTokenAuthenticationPolicy({
+      credential: { getToken: getTokenStub },
+      scopes: ["https://example.org"],
+      challengeCallbacks: {
+        authorizeRequestOnChallenge: storageAuthorizeRequestOnChallenge,
+      },
+    });
+
+    const sendTestRequest = () =>
+      policy.sendRequest(
+        {
+          headers: createHttpHeaders(),
+          method: "GET",
+          requestId: "",
+          timeout: 1000,
+          url: "https://example.org",
+          withCredentials: true,
+        },
+        async (req) => {
+          assert.equal(req.headers.get("authorization"), "Bearer originalToken");
+          return {
+            headers: createHttpHeaders(),
+            request: req,
+            status: 200,
+          };
+        }
+      );
+
+    // Send the request 2 times
+    await sendTestRequest();
+    await sendTestRequest();
+
+    // We are setting expiration time within the refresh window
+    // so we expect 2 calls to refresh the token
+    assert.equal(getTokenStub.callCount, 2);
+  });
+});
```

Comment:
Should there be unit tests for the individual new functions used underneath?

Reply:
Those are not exported, however, let me see if I can add some other tests that exercise more of the functions underneath.